Added ability to add custom port on pod and service port forwarding

rdeepc opened the pull request:
Signed-off-by: rdeepc <12953177+rdeepc@users.noreply.github.com>

rdeepc committed:
Added ability to add custom port on pod and service port forwarding

Signed-off-by: rdeepc <12953177+rdeepc@users.noreply.github.com>

rdeepc committed:
- Added Custom Port Selection for port forwarding
- Implemented Random Port if custom port is not provided

Signed-off-by: rdeepc <12953177+rdeepc@users.noreply.github.com>

rdeepc committed:
Added ability to add custom port on pod and service port forwarding

Signed-off-by: rdeepc <12953177+rdeepc@users.noreply.github.com>

rdeepc committed:
Merge branch 'lensapp:master' into feature/port-forward-custom-port-input

Nokel81 commented on src/main/routes/port-forward-route.ts:
```diff
@@ -44,10 +45,10 @@ class PortForward {
+        pf.kind == forward.kind &&
+        pf.name == forward.name &&
+        pf.namespace == forward.namespace &&
+        pf.port == forward.port
```
Please revert formatting changes.

Nokel81 commented on src/main/routes/port-forward-route.ts:
```diff
@@ -74,7 +77,7 @@ class PortForward {
       "-n", this.namespace,
       `${this.kind}/${this.name}`,
-      `:${this.port}`
+      `${this.forwardPort??""}:${this.port}`
```
If `forwardPort` can be of type `null | undefined` then the fields should all be set to either `optional` or include those types. (I think optional would be better)

Nokel81 commented on src/main/routes/port-forward-route.ts:
```diff
@@ -118,33 +125,73 @@ class PortForward {
-    respondJson(response, {});
+    portForward.stop().then( () => {
+      respondJson(response, {status: true});
+    }).catch(error => logger.error(error));
```
Please do not use promise functions in an async context.

Nokel81 commented on src/main/routes/port-forward-route.ts:
```diff
@@ -118,33 +125,73 @@ class PortForward {
-      }
+    if (portForward) {
+      forwardedPort = Number(portForward.forwardPort);
     }
```
This can be simplified:

```suggestion
    const forwardedPort = Number(portForward?.forwardPort ?? -1);
```

Nokel81 commented on src/renderer/components/+network-services/service-port-component.tsx:
```diff
@@ -28,20 +28,44 @@ import { apiBase } from "../../api";
 @observer
 export class ServicePortComponent extends React.Component<Props> {
   @observable waiting = false;
+  @observable forwardPort = -1;
```
I think it would be better to use `null` or `undefined` as the sentinel value instead of a special number.

Nokel81 commented on src/renderer/components/+network-services/service-port-component.tsx:
```diff
@@ -28,20 +28,44 @@ import { apiBase } from "../../api";
+  }
+
+  init() {
+    this.checkExistingPortForwarding().then();
```
Please remove the `then`. If you are meaning to ignore the output, either do `.catch(noop)` or match `init()` an async function, await the result in a `try/catch` and then either log the error or add a comment indicating that you want to ignore it. Best case would be to print an error message.

Nokel81 commented on src/renderer/components/+network-services/service-port-component.tsx:
```diff
@@ -61,14 +104,32 @@ export class ServicePortComponent extends React.Component<Props> {
+          disabled={this.isPortForwarded}
+          onChange={(value) => this.forwardPort = Number(value)}
+        />
+        <Button onClick={() => portForwardAction()}> {this.isPortForwarded ? "Stop":"Forward"} </Button>
```
Please add some screenshots of how this looks

rdeepc replied:
![2021-07-09_23-37](https://user-images.githubusercontent.com/12953177/125150574-c1136e80-e10e-11eb-99bf-99cdde11276b.png)


![2021-07-09_23-39](https://user-images.githubusercontent.com/12953177/125150597-fd46cf00-e10e-11eb-9eec-57f07bdde740.png)

rdeepc replied:
Thanks for the feedback, I have made the changes as suggested

rdeepc committed:
Merge branch 'lensapp:master' into feature/port-forward-custom-port-input

rdeepc committed:
Added ability to add custom port on pod and service port forwarding

- pod-container-port fixed init async and active port checking
- service-port-component fixed init async and active port checking
- port-forward-route promise and response fix

Signed-off-by: rdeepc <12953177+rdeepc@users.noreply.github.com>

rdeepc committed:
Merge branch 'master' into feature/port-forward-custom-port-input

jim-docker commented on src/renderer/components/+network-services/service-port-component.tsx:
```diff
@@ -61,14 +105,32 @@ export class ServicePortComponent extends React.Component<Props> {
     const { port } = this.props;
 
+    if (this.forwardPort == null) {
+      this.forwardPort = port.port;
```
I think this is generally not a good idea. If the pod or service port is 80 then suggesting the forward be 80 is a problem. Also, this will cause problems for frequently used ports like 8080, when you want to forward more than one service that exposes 8080. 
If `this.forwardPort` is `null` I think the port input field should start as blank (or some placeholder text like "random"  or "arbitrary port"). If the user leaves it blank then Lens does what it did before (lets `kubectl` find a port)

rdeepc replied:
Thanks for the suggestion I have added the random port option if a port is not given, with a Random placeholder

jim-docker commented on src/renderer/components/+network-services/service-port-component.tsx:
```diff
@@ -61,14 +105,32 @@ export class ServicePortComponent extends React.Component<Props> {
+        <text>to</text>
+        <Input className={"portInput"}
+          type="number"
+          min="0"
```
```suggestion
          min="1"
```
I think for TCP, port number 0 is reserved and cannot be used. I guess it's debatable if any of the low numbers should be used

rdeepc replied:
I have used the min port as 0, as kubectl allocates random port for input 0, Thanks for your input

jim-docker reviewed (commented):
thanks for contributing this!

jim-docker reviewed (approved):
@rdeepc I've changed the target for this PR to `lensapp:feature/port-forward-dashboard` (which is currently even with `master`). In July I used your work as a starting point for a port-forward dashboard that I implemented in a hackweek project here at Mirantis (see https://github.com/jim-docker/lens/tree/port-forward-dashboard). I am now beginning a proper integration into Lens, starting with your work, which will eventually get merged with `master`.  Your PR will require some changes to work with the dashboard, and I assume you are OK with us continuing the work? Thanks again for this contribution
